### PR TITLE
feat(ui): [Feature] Bidirectional URL query params sync for list block filters #764

### LINKED ISSUE #764: [Feature] Bidirectional URL query params sync for list block filters
**Is your feature request related to a problem? Please describe.**

List blocks (ticket-list, order-list, invoice-list, product-list, notification-list) manage filter state locally via `useState` + Formik. When a user applies filters, the URL does not update — meaning filtered views cannot be shared, bookmarked, or linked to from other parts of the application (e.g. clickable status boxes navigating to a pre-filtered list).

**Describe the solution you'd like**

Implement bidirectional synchronization between filter state and URL query parameters across all list blocks.

**Key requirements:**
- New `useUrlFilters` hook in `packages/ui/src/hooks/` that replaces `useState(initialFilters)` in list blocks
- Serialization utilities for converting filter state to/from URL query params
- Namespace-per-block prefixing (`{ns}_key=value`) to avoid collisions when multiple blocks appear on the same page
- Repeated keys for multi-select values (e.g. `status=open&status=pending`)
- Pagination as `{ns}_page=N` (1-based, human-readable)
- View mode in URL (`{ns}_view=grid`), omitted when default (`list`)
- Only non-default filter values written to URL (clean URLs)
- `router.replace` for filter changes (no browser history pollution)
- Dependency injection pattern — hook accepts `searchParams` and `onUrlChange` as arguments (no direct `next/navigation` imports), making it framework-agnostic

**Blocks to migrate:**
1. `ticket-list` (proof of concept)
2. `order-list`
3. `invoice-list`
4. `product-list`
5. `notification-list`

**Benefits:**
- Shareable/bookmarkable filtered views
- Clickable UI elements (e.g. status summary boxes) can navigate to pre-filtered lists
- Backward compatible — blocks without the hook continue working as before

**Additional context (if applicable)**

### Implementation approach

**New hook: `useUrlFilters`** (`packages/ui/src/hooks/use-url-filters.ts`)

Replaces `useState(initialFilters)` in list blocks. Uses dependency injection — accepts `searchParams` and `onUrlChange` as arguments instead of importing from `next/navigation` directly:

```typescript
interface UseUrlFiltersOptions<TFilters extends Record<string, string | number | string[]>> {
    initialFilters: TFilters;
    namespace?: string;         // prefix for URL params, e.g. 'ticket'
    excludeKeys?: (keyof TFilters)[];
    searchParams: URLSearchParams;
    onUrlChange: (params: string) => void;
}
```

**Serialization rules:**
- Simple values: `ticket_status=open`
- Multi-value: `ticket_status=open&ticket_status=pending` (repeated keys)
- Pagination: `ticket_page=2` (1-based, derived from offset/limit)
- View mode: `ticket_view=grid` (omitted when default `list`)
- Only non-default values written to URL

**New utility functions** (`packages/ui/src/hooks/use-url-filters.utils.ts`):
- `serializeFiltersToParams()` — filter state → URL query string
- `deserializeParamsToFilters()` — URL query string → filter state
- `filtersToPage()` / `pageToOffset()` — pagination helpers

### Files to create
- `packages/ui/src/hooks/use-url-filters.ts` — the hook
- `packages/ui/src/hooks/use-url-filters.utils.ts` — serialization/deserialization utilities
- `packages/ui/src/hooks/use-url-filters.test.ts` — unit tests

### Files to modify
- `packages/blocks/ticket-list/src/frontend/TicketList.client.tsx`
- `packages/blocks/order-list/src/frontend/OrderList.client.tsx`
- `packages/blocks/invoice-list/src/frontend/InvoiceList.client.tsx`
- `packages/blocks/product-list/src/frontend/ProductList.client.tsx`
- `packages/blocks/notification-list/src/frontend/NotificationList.client.tsx`

### Verification checklist
- [ ] Filter change updates URL query params
- [ ] Pasting URL with filters applies them on mount
- [ ] Reset clears URL params
- [ ] Pagination reflected in URL (`?ticket_page=2`)
- [ ] View mode reflected in URL (`?ticket_view=grid`)
- [ ] Two blocks on the same page use separate namespaces without collisions
<br/>
<hr/>

<details><summary>This repo is using Opire - what does it mean? 👇</summary><br/>💵 Everyone can add rewards for this issue commenting <code>/reward 100</code> (replace <code>100</code> with the amount).<br/>🕵️‍♂️ If someone starts working on this issue to earn the rewards, they can comment <code>/try</code> to let everyone know!<br/>🙌 And when they open the PR, they can comment <code>/claim #764</code> either in the PR description or in a PR's comment.<br/><br/>🪙 Also, everyone can tip any user commenting <code>/tip 20 @michnowak</code> (replace <code>20</code> with the amount, and <code>@michnowak</code> with the user to tip).<br/><br/>📖 If you want to learn more, check out our <a href="https://docs.opire.dev">documentation</a>.</details>

### PR DESCRIPTION
**Pull request by lewanp**
**What does this PR do?**

- [x] My feature

**Related Ticket(s)**

- Closes #764 — [Feature] Bidirectional URL query params sync for list block filters

**Key Changes**

List block filter state lives in the URL, so a filtered view can be shared, bookmarked and linked to. The URL is handled at both ends of the render: the **first response is filtered on the server**, every **later change stays on the client** and only rewrites the address bar.

**The hook and the URL contract** — `packages/ui/src/hooks/use-url-filters.ts` + `use-url-filters.utils.ts`

- `useUrlFilters` replaces `useState(initialFilters)` in the list blocks and returns `filters`, `setFilters`, `resetFilters`, `viewMode`, `setViewMode`, `isRestoredFromUrl`. Reading and writing the URL is injected (`searchParams`, `onUrlChange`), so `@o2s/ui` keeps no dependency on `next/navigation`.
- Filter changes are written with the **History API** (`replaceUrlParams`), not `router.replace`: the block fetches its own data, so a filter click must not trigger an RSC navigation that re-renders the whole route. No history entries are added.
- Two param conventions, one per block: namespaced (`?ticket_status=OPEN&ticket_page=2`) so several list blocks can share a page, or plain (`?category=TOOLS&page=2`) for a public, indexed list, which passes `filterKeys` instead of a `namespace`. Pagination is a 1-based `page`, view mode is `view`, and only values differing from the block defaults are written.
- Writes merge into the **live** query string, and each setter derives the next state from the latest one, so two blocks writing in turn — or a filter change and a view-mode change in one batch — cannot drop each other's params.

**Server-rendered filters** — `apps/frontend`, all five list blocks

- `searchParams` travel from the page through `PageTemplate` → templates → `renderBlocks` into the blocks (`BlockSearchParams` on `BaseBlockProps`, optional, so blocks that ignore it are unaffected). `parseFiltersFromSearchParams` in `@o2s/ui` turns them into a block query, next to the hook that writes them.
- Each block query takes a 1-based `page` that the **API** resolves into an `offset` using the page size from the CMS block config — only the API knows that number.
- With the server rendering the filtered state, the client's refetch on mount is gone: a filtered link costs one request instead of two, and no unfiltered list flashes first.

**SEO for the public product list** — `apps/frontend/src/utils/seo.ts`

| URL | Canonical | Robots |
| --- | --- | --- |
| `/products` | `/products` | `index, follow` |
| `/products?category=TOOLS` | `/products?category=TOOLS` | `index, follow` |
| `/products?category=TOOLS&page=3` | `/products` | `noindex, follow` |
| `/products?sort=price_asc` | `/products` | `noindex, follow` |

One value of one whitelisted facet (`INDEXABLE_FILTERS`) still describes a page worth indexing; sorting, deep pages and facet combinations canonicalise back. Facet values are also rendered as real `<a href>` links, because a crawler follows links but never operates a select — and following one rebuilds the block from the server data for those params (`searchParamsKey`).

**Problems found while verifying, each fixed here**

- **Multi-select filters must be restored as arrays** — `?ticket_status=CLOSED` deserialised to a string, so the toggle group iterated it character by character and the next click produced `ticket_status=C&ticket_status=L&…`. The blocks derive `multiValueKeys` from the CMS config, and only for toggle groups: a select writes a single string back whatever `allowMultiple` says, and handing it an array tripped React's `<select>` check.
- **Reset has to clear CMS-driven keys** — `status`, `sort`, `topic` have no entry in `initialFilters`, so with a namespace every prefixed param is rewritten from scratch, and without one the block names its keys.
- **Filtering did not reset pagination** — changing a filter on page 3 kept the offset and showed page 3 of a different, usually shorter result set. Paging moved to its own handler so an explicit offset still survives.
- **Filters became unreachable on an empty result** — the controls were hidden when the *server* data was empty, which after server-side filtering meant filtering down to nothing hid the only way to undo it.
- **The active-filter counter** compared values by reference and counted an empty array as active, so the drawer variant showed "Remove filters (1)" with nothing selected.
- **Odd pagination params** — an explicit `offset=0` lost to a `page` further down the query string, and `?page=2.5` or `?page=Infinity` travelled on as a fractional or infinite offset (a silently empty list here, a rejected query on a real backend). Only a value that can be one resolves now, in the parser, in `pageToOffset` and in the API.
- **`searchParams` did not reach nested blocks** — a list block configured inside a `CategoryBlock` fell back to the defaults with nothing left to correct it.
- **The mocked tickets integration ignored the array form** its own contract documents, so any multi-status selection returned nothing.
- **A failed product list refetch** rejected unhandled instead of showing the request-error toast.

**Tests**

- `packages/ui` had no test project. It gets one (a `ui` variant of the shared vitest config, node environment) with **37 tests** over the serialisation helpers, the search-param parser and the address-bar helpers — 97% of the module's statements.
- The hook itself needs rendering, and neither `jsdom` nor `@testing-library/react` is installed, so **9 tests** run in the browser project the Storybook tests already use (React 19 `act`, Playwright browsers are installed by CI anyway). They pin down the batched-setter and live-URL behaviours above.
- Each block's API service gets a spec for the page resolution: CMS page size, page as a number and as a query string, page 1, an explicit offset winning, `'nonsense'` / `2.5` / `Infinity` / `MAX_VALUE` staying on the first page, no `page` leaking into the domain query, filters passed through.

**Side effects and deliberate limits**

- Filter changes add no history entries, so Back leaves the list rather than undoing the last filter; links — the facet links included — are real entries that Back and Forward move across correctly. Documented in `apps/docs/.../routing.md`.
- The blocks' query params stay namespaced except the product list, which is public and indexed and therefore plain.
- `renderBlocks` and the `Category` block's `renderBlocks` prop take an extra optional argument.
- Multi-value filtering is limited by each module's contract: the tickets module takes repeated `status` values, the others take one value per filter, so a repeated param contributes its first. Widening that would mean changing the module contracts across all integrations.
- Blocks that do not use the hook are untouched.

**How to test**

No migrations or new dependencies needed. `npm run predev && npm run dev`, log in as `jane@example.com` / `admin`.

1. **Filtering without a page reload** — on **Cases** change a status filter. The URL becomes `?ticket_status=OPEN`; DevTools shows a single `GET /api/blocks/ticket-list?…` and **no document or `_rsc` request**. `history.length` does not grow.
2. **A filtered link renders filtered** — open `/en/cases?ticket_status=OPEN&ticket_status=CLOSED&ticket_status=IN_PROGRESS&ticket_topic=CONTACT_US&ticket_sort=status_DESC` in a new tab. The rows are Contact Form only, sorted by status, the toggles are restored — and there is **no client-side API request at all**, the first response already carries it.
3. **Pagination in the URL** — append `&ticket_page=2`; the pager reads "2 of 3 pages" server-side. Then change a filter: the page resets to 1 and `ticket_page` disappears.
4. **Namespace isolation** — on a page with two list blocks, change one block's filter and confirm the other's params plus unrelated ones (`?tab=details`) survive.
5. **Public product list** — `/en/products?category=TOOLS` returns 12 filtered products with `<link rel="canonical" href=".../products?category=TOOLS">` and `robots: index, follow`; adding `&page=2` or `&sort=…` switches the canonical to `/products` and `robots` to `noindex, follow`. The category links at the bottom of the block are real links: click one and the list, the URL and the active facet move together; Back and Forward restore both.
6. **Empty result stays usable** — `/en/products?category=SOFTWARE` (no matches) still renders the filter controls, and switching the category recovers results. Same for `/en/cases?ticket_status=ALL`.
7. **Odd params** — `?page=2.5`, `?page=Infinity` and `?page=-2` render the first page instead of an empty list.
8. **Tests** — `npm test`, or `npx vitest run` in `packages/ui` (37) and `npx vitest run --project=hooks` from the root (9).

Verified in a browser against a running frontend and api-harmonization:

| Check | Result |
| --- | --- |
| Filter change: requests | 1× API, 0× document/RSC; `history.length` unchanged |
| Filtered link, 3 statuses + topic + sort | server-rendered, 0× client API request |
| `ticket_page=2` on top of those filters | "2 of 3 pages", other 10 IDs |
| `/en/invoices?invoice_paymentStatus=PAYMENT_COMPLETE&invoice_sort=issuedDate_ASC&invoice_page=2` | 5 rows, all Paid, "2 of 6 pages", 0× client API request |
| `/en/orders?order_status=COMPLETED&order_sort=createdAt_ASC` | 10 rows, all Completed, "1 of 9 pages" |
| `/en/notifications?notification_priority=HIGH&…&notification_page=2` | 10 rows, all High, "2 of 4 pages" |
| `/en/products?category=MEASUREMENT` | 1 product, self-canonical, `index, follow` |
| Facet link click, then Back / Forward | 12 ↔ 1 product, active facet follows |
| `?category=SOFTWARE` (0 matches) | filter form present, switching category recovers 12 products |
| `?page=2.5` / `Infinity` / `-2` | first page, no fractional offset reaches the API |
| Two namespaces on one page | foreign and page params preserved |

`tsc --noEmit` and `eslint --max-warnings=0` are clean for every touched package, `npm test` passes (46 new tests in `@o2s/ui`, 28 across the five block services), and the list block stories still render.

**Media (Loom or gif)**

- N/A

## Summary by CodeRabbit

* **New Features**
  * Added URL-synchronized filters across ticket, order, invoice, product, and notification lists.
  * Filter selections, multi-select values, pagination, and view modes are restored from shareable URLs.
  * Added server-rendered filtering, SEO-friendly category links, and canonical URL handling.
  * Filter updates preserve unrelated URL parameters without full-page navigation.
  * Filters and empty states remain accessible while lists load or contain no results.

* **Bug Fixes**
  * Filter changes reset pagination to the first page.
  * Improved multi-value filter counts and select restoration.
  * Product refetch failures now display an error notification.
  * Batched updates no longer overwrite newer URL parameters.

### CODE REVIEW

File: packages/ui/src/hooks/use-url-filters.ts [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+ * Keeps list block filter state in sync with the URL query string.
+ *
+ * A drop-in replacement for `useState(initialFilters)`: state is seeded from the URL on mount and
+ * every change is pushed back into it. Navigation is injected rather than imported, so the hook
+ * stays free of `next/navigation` and can be used (and tested) outside a Next.js tree.
+ *
+ * The URL is read once, on mount. Filter changes are expected to be written with `router.replace`,
+ * which adds no history entries, so there is no later URL change to read back.
+ */
+export const useUrlFilters = <TFilters extends object>({
```
Comment (marcinkrasowski) [inline packages/ui/src/hooks/use-url-filters.ts:63]: please check if we can add unit tests for this, as well for the added utils

File: packages/blocks/notifications/notification-list/src/api-harmonization/notification-list.service.ts [hunk 2]
```diff
@@ -11,6 +11,35 @@ import { GetNotificationListBlockQuery } from './notification-list.request';
...
+/** The page size the block renders with: an explicit query value, then the CMS config, then the default. */
+const resolveLimit = (query: GetNotificationListBlockQuery, cmsLimit?: number): number =>
+    Number(query.limit) || Number(cmsLimit) || DEFAULT_LIMIT;
+
+/**
+ * `offset` wins whenever it is given, `0` included; otherwise a 1-based `page` is resolved with the
+ * page size above, which is why this lives here and not in the caller: only the API knows the CMS
+ * pagination config.
+ */
+const resolveOffset = (query: GetNotificationListBlockQuery, limit: number): number => {
```
Comment (marcinkrasowski) [inline packages/blocks/notifications/notification-list/src/api-harmonization/notification-list.service.ts]: can/should we extract these to some reusable utils? looks like it is repeated in every block

Comment (marcinkrasowski) [reply packages/blocks/notifications/notification-list/src/api-harmonization/notification-list.service.ts]: and instead of `GetNotificationListBlockQuery` we could use `PaginationQuery` which would drop block-level type requirement

File: packages/blocks/orders/order-list/src/frontend/OrderList.client.tsx [hunk 3]
```diff
@@ -42,54 +44,90 @@ export const OrderListPure: React.FC<OrderListPureProps> = ({ locale, accessToke
...
     })();
 
     const [data, setData] = useState<Model.OrderListBlock>(component);
-    const [filters, setFilters] = useState(initialFilters);
-    const [viewMode, setViewMode] = useState<'list' | 'grid'>(initialViewMode);
     const [selectedRows, setSelectedRows] = useState<Set<string | number>>(new Set());
 
     const [isPending, startTransition] = useTransition();
 
+    const pathname = usePathname();
```
Comment (marcinkrasowski) [inline packages/blocks/orders/order-list/src/frontend/OrderList.client.tsx:51]: here as well, the same/similar code appears in every block - can we extract at least parts of it to utils? we have `packages/utils/api-harmonization` and `packages/utils/frontend` which are exactly for the cases like this

**Review by marcinkrasowski — COMMENTED** [review]
a few more things on top of the extraction points i already left inline.

most of it is the same theme, the five list blocks are almost identical and could share more:
- the backend `resolveOffset`/`resolveLimit` (my other comment) is byte-identical in all five services, only the query type differs. good candidate for `@o2s/utils.api-harmonization`, typed on `PaginationQuery`.
- the client wiring (`pathname`/`searchParams`/`handleUrlChange`/`multiValueKeys`/`handleFilter`/`handlePageChange`/`handleReset`) is ~60 near-identical lines per block. a small `useListBlockUrlFilters` in `@o2s/utils.frontend` would cover most of it.
- `page` and its jsdoc are copy-pasted into every `*.request.ts`. if the block query extended `PaginationQuery` it'd live in one place.
- the renderers are also near-identical, a shared list renderer taking id/searchParams/fallback would remove them.
- the page to offset rule now lives in three spots (`pageToOffset`, `parseFiltersFromSearchParams`, and each `resolveOffset`) and they all have to stay in agreement.

smaller stuff:
- server/renderer naming drifts between blocks (`TicketListServer`/`TicketListRenderer` vs `OrderList`/`Renderer`). since this touches all of them it'd be a good moment to align.
- the three list service specs are ~120+ lines each and mostly parallel. if the resolver moves to a util its pagination cases get tested once and the specs shrink.

a few seo ones are inline too (public/private pages, tracking params, the duplicated facet list, and no tests on seo.ts yet).

also the hook/utils tests i asked about earlier look like they landed in bace6218, so that thread can probably be resolved.

File: apps/frontend/src/utils/seo.ts [hunk 4]
```diff
@@ -8,6 +8,39 @@ import { DEFAULT_LOCALE, SUPPORTED_LOCALES } from '@/i18n/routing';
...
+
+const getFilterSeo = (searchParams: SearchParams = {}): FilterSeo => {
+    const active = Object.entries(searchParams).filter(([, value]) => value !== undefined && value !== '');
+
+    if (!active.length) {
+        return { canonicalQuery: '', noIndex: false };
+    }
+
+    const [key, value] = active[0]!;
+    const isSingleFacet = active.length === 1 && INDEXABLE_FILTERS.includes(key) && typeof value === 'string';
```
Comment (marcinkrasowski) [inline apps/frontend/src/utils/seo.ts]: getFilterSeo treats any query param as a filter here, so a url with `utm_source`/`gclid` (or anything not in `INDEXABLE_FILTERS`) makes the page `noindex` and drops the canonical back to the bare url. `?category=TOOLS&utm_source=x` loses its `?category=TOOLS` canonical too. can we key this off the allowlist and just ignore unknown params? and like the filter utils, worth covering these cases with unit tests, there are none for seo.ts yet.

File: apps/frontend/src/utils/seo.ts [hunk 5]
```diff
@@ -34,21 +69,24 @@ export const generateSeo = ({
...
+    const filters = getFilterSeo(searchParams);
+    const canonical = `${url}${filters.canonicalQuery}`;
 
     return {
         title,
         description,
         keywords,
         robots: {
-            index: !noIndex,
+            index: !noIndex && !filters.noIndex,
```
Comment (marcinkrasowski) [inline apps/frontend/src/utils/seo.ts:92]: one case that seems missing: public vs private pages. `/cases` needs auth but the bare url still comes back `index, follow` (only the filtered variants go noindex, and only because they carry params). should authenticated pages be forced `noindex` here regardless of filters, e.g. off a page-level flag? right now it leans on each page setting `seo.noIndex` by hand and the mocks have it `false`.

File: apps/frontend/src/utils/seo.ts [hunk 6]
```diff
@@ -8,6 +8,39 @@ import { DEFAULT_LOCALE, SUPPORTED_LOCALES } from '@/i18n/routing';
...
 
+type SearchParams = Record<string, string | string[] | undefined>;
+
+/**
+ * Filter params whose single-value form still describes a page worth indexing on its own — a category
+ * listing is a page, a sort order or a deep page of one is a variant of it. Everything else (`page`,
+ * `sort`, `view`, several values of one facet, several facets at once) canonicalises back and is kept
+ * out of the index, so filtering cannot spray near-duplicates across the crawl budget.
+ */
+const INDEXABLE_FILTERS = ['category'];
```
Comment (marcinkrasowski) [inline apps/frontend/src/utils/seo.ts]: `INDEXABLE_FILTERS` here and `SEO_FACETS` in `ProductList.client.tsx` are the same list kept in sync by a comment. can we move it to one shared const so they can't drift?

File: packages/ui/src/components/Forms/Filters/FiltersContext.tsx [hunk 7]
```diff
@@ -12,30 +14,44 @@ type FiltersContextType = {
 
 const FiltersContext = createContext<FiltersContextType | null>(null);
 
+/** Keys describing the block rather than a user filter. */
+const EXCLUDED_KEYS = ['offset', 'limit', 'id'];
```
Comment (marcinkrasowski) [inline packages/ui/src/components/Forms/Filters/FiltersContext.tsx]: these excluded keys are also defined in `use-url-filters.utils.ts` as `DEFAULT_EXCLUDED_KEYS`, and the two lists aren't the same (`offset` is only here). same idea in two places, can we share one const?

File: packages/blocks/support/ticket-list/src/frontend/TicketList.server.tsx [hunk 8]
```diff
@@ -15,13 +25,19 @@ export const TicketListServer: React.FC<TicketListProps> = async ({
...
     try {
         data = await sdk.blocks.getTicketList(
             {
                 id,
                 preview: isDraftModeEnabled,
+                ...(parseFiltersFromSearchParams(searchParams, {
+                    namespace: NAMESPACE,
+                    keys: FILTER_KEYS,
+                    multiValueKeys: MULTI_VALUE_KEYS,
+                }) as Partial<Request.GetTicketListBlockQuery>),
```
Comment (marcinkrasowski) [inline packages/blocks/support/ticket-list/src/frontend/TicketList.server.tsx]: this `as Partial<...>` cast repeats in every block's server. could `parseFiltersFromSearchParams` be generic over the query type so we can drop the cast in all of them?

File: packages/blocks/notifications/notification-list/src/api-harmonization/notification-list.service.ts [hunk 9]
```diff
@@ -11,6 +11,35 @@ import { GetNotificationListBlockQuery } from './notification-list.request';
 
 const H = HeaderName;
 
+const DEFAULT_LIMIT = 1;
```
Comment (marcinkrasowski) [inline packages/blocks/notifications/notification-list/src/api-harmonization/notification-list.service.ts]: is `DEFAULT_LIMIT = 1` intended? one row feels like an odd fallback page size. it's also repeated in every service, so it'd move together with the resolver extraction above.

File: packages/framework/src/utils/models/pagination.ts [hunk 10]
```diff
@@ -3,6 +3,63 @@ export class PaginationQuery {
...
+
+export interface ResolvePaginationOptions {
+    /** Page size from the block's CMS config, when it has one. */
+    cmsLimit?: number;
+    /** Page size to use when neither the query nor the CMS config names one. */
+    defaultLimit: number;
+}
+
+/** A page size only counts as one when it is a whole number of rows above zero. */
+const toLimit = (value: number | undefined): number | undefined => {
```
Comment (marcinkrasowski) [inline packages/framework/src/utils/models/pagination.ts]: please move it to a dedicated utils package - `packages/framework/src/utils/models` should only contain models (TS interfaces/types) and not any helper functions; the correct place for this would be `packages/utils/frontend/src/utils` that can be then used on the frontend like:
```
import { Utils } from '@o2s/utils.frontend';
...
Utils.FormatAddress.formatStreetAddress();
```

**Review by marcinkrasowski — APPROVED** [review]
